File: java/kata/src/main/java/necklaceRefactoringKata/Packer.java
package necklaceRefactoringKata;

import necklaceRefactoringKata.jewellery.Earring;
import necklaceRefactoringKata.jewellery.EarringType;
import necklaceRefactoringKata.jewellery.Jewel;
import necklaceRefactoringKata.jewellery.Jewellery;
import necklaceRefactoringKata.jewellery.Necklace;
import necklaceRefactoringKata.jewellery.PendantNecklace;

public class Packer {

    public static void packNecklace(Necklace item, JewelleryStorage storage) {
        if (item.stone == Jewel.Diamond) {
            storage.safe.add(item);
        } else if (!item.isHeavy()) {
            storage.box.topShelf.add(item);
        } else if (item instanceof PendantNecklace pendantNecklace) {
            storage.tree.add(pendantNecklace.chain);
            storage.box.topShelf.add(pendantNecklace.pendant);
        } else {
            storage.tree.add(item);
        }
    }

    public static void pack(Jewellery item, JewelleryStorage storage) {
        if (storage.IsInTravelRoll(item) && !item.isHeavy()) {
            storage.box.topShelf.add(item);
        } else if (item.stone == Jewel.Diamond) {
            storage.safe.add(item);
        } else if (item instanceof PendantNecklace pendantNecklace) {
            storage.tree.add(pendantNecklace.chain);
            storage.box.topShelf.add(pendantNecklace.pendant);
        } else if (item.isSmall()) {
            storage.box.topShelf.add(item);
        } else if (!item.isHeavy() && item instanceof Necklace) {
            storage.box.topShelf.add(item);
        } else if (item instanceof Earring earring && earring.type == EarringType.Hoop) {
            storage.tree.add(earring);
        } else if (item instanceof Earring earring && earring.type == EarringType.Drop && earring.stone != Jewel.Plain) {
            storage.box.topShelf.add(earring);
        } else if (item instanceof Earring earring && earring.type == EarringType.Drop) {
            storage.box.mainSection.add(earring);
        } else if (item instanceof Necklace necklace) {
            storage.tree.add(necklace);
        } else {
            storage.dresserTop.add(item);
        }

        if (storage.IsInTravelRoll(item))
            storage.travelRoll.remove(item);
    }
}

File: kotlin/src/main/kotlin/org/sammancoaching/necklace/Packer.kt
package org.sammancoaching.necklace

fun packNecklace(item: Necklace, storage: JewelleryStorage) {
  if (item.stone == Jewel.Diamond) {
    storage.safe.add(item)
  } else if (item.isLarge().not()) {
    storage.box.topShelf.add(item)
  } else if (item is PendantNecklace) {
    storage.tree.add(item.chain)
    storage.box.topShelf.add(item.pendant)
  } else {
    storage.tree.add(item)
  }
}

fun pack(item: Jewellery, storage: JewelleryStorage) {
  if (storage.isInTravelRoll(item) and item.isLarge().not()) {
    storage.box.topShelf.add(item)
  } else if (item.stone == Jewel.Diamond) {
    storage.safe.add(item)
  } else if (item.isSmall()) {
    storage.box.topShelf.add(item)
  } else if (item is Earring && item.type == EarringType.Hoop) {
    storage.tree.add(item)
  } else if (item is Earring && item.type == EarringType.Drop && item.stone != Jewel.Plain) {
    storage.box.topShelf.add(item)
  } else if (item is Earring && item.type == EarringType.Drop) {
    storage.box.mainSection.add(item)
  } else if (item is PendantNecklace && item.type == NecklaceType.Pendant) {
    storage.tree.add(item.chain)
    storage.box.topShelf.add(item.pendant)
  } else if (item.isNecklace()) {
    storage.tree.add(item)
  } else {
    storage.dresserTop.add(item)
  }

  if (storage.isInTravelRoll(item)) {
    storage.travelRoll.remove(item)
  }
}
